### PR TITLE
[CI] Skip irrelevant checks

### LINKED ISSUE #483: [CI] Disable checks from other crates
### Which component is your feature related to?

Cast

### Feature Request

Forge tests could be omitted on pull requests with the `Cast` label, or if the files (the hash of files maybe?) from other crates are left unchanged.

### PR DESCRIPTION
**Pull request by drknzz**
Closes #483 

## Introduced changes

- Detect changes in crates (cast, forge, cheatnet) and run tests only if their files have changed
- Changes outside of the crates count as global changes, resulting in all tests being ran

## Checklist

- [x] Linked relevant issue
- [x] Updated relevant documentation
- [x] Added relevant tests
- [x] Performed self-review of the code
- [x] Added changes to `CHANGELOG.md`

### CODE REVIEW

**Review by THenry14 — APPROVED** [review]
LGTM! let's wait for someone from forge and cheatnet though, but this is lit 🔥

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -134,3 +143,26 @@ jobs:
...
+    runs-on: ubuntu-latest
+    permissions:
+      pull-requests: read
+    outputs:
+      cast: ${{ steps.filter.outputs.cast }}
+      cheatnet: ${{ steps.filter.outputs.cheatnet }}
+      forge: ${{ steps.filter.outputs.forge }}
+      global: ${{ steps.filter.outputs.global }}
+    steps:
+    - uses: dorny/paths-filter@v2
```
Comment (cptartur) [inline .github/workflows/ci.yml:157]: I think we can fix a specific commit here instead of just `@v2`.

**Review by cptartur — COMMENTED** [review]
I wonder if we should be always running all checks on master branch (and perhaps in the merge queue)? I see the benefit of only checking related changes in the PR itself, but I worry that always running checks conditional we'll make us miss some errors.

**Comment by drknzz** [comment]
> I wonder if we should be always running all checks on master branch (and perhaps in the merge queue)? I see the benefit of only checking related changes in the PR itself, but I worry that always running checks conditional we'll make us miss some errors.

Unless somehow changes from one crate would impact another one (in which case it would be some giant flaw in the repo structure) it's safe to run only conditional checks, but I'm down with the idea of running the full suite on merged PRs (as it does not slow down the development process anyway).

**Comment by cptartur** [comment]
> > I wonder if we should be always running all checks on master branch (and perhaps in the merge queue)? I see the benefit of only checking related changes in the PR itself, but I worry that always running checks conditional we'll make us miss some errors.
> 
> Unless somehow changes from one crate would impact another one (in which case it would be some giant flaw in the repo structure) it's safe to run only conditional checks, but I'm down with the idea of running the full suite on merged PRs (as it does not slow down the development process anyway).

The split is cleared with cast and forge, but in case of cheatnet there very much can be a case where some part of the code gets changed there and it starts affecting the behavior of forge as well.

**Comment by drknzz** [comment]
> very much can be a case

Why so?
What kind of changes to cheatnet crate would affect forge crate?

**Comment by drknzz** [comment]
Maybe we should just leave the cast crate separation from other crates then?

**Comment by THenry14** [comment]
It may be a good idea to run all the checks for master, as Artur suggested :+1: this shouldn't affect anyone, and could provide us with additional layer of 'security' - compatibility wise. It also shouldn't be too hard to accomplish.

**Comment by Arcticae** [comment]
I would also run all checks on merge queue as well as on master, so we don't have to revert the merges

**Comment by cptartur** [comment]
Wonder if this should still be included considering that #505 introduces a considerate speedup anyway.

**Pull request closed without merging**